### PR TITLE
Fix timezone issue segment rebuild

### LINKED ISSUE #7967: Relative date segment filter (like -3 days/+3 days) doesn't take into account Mautic timezone setting
[//]: # ( Invisible comment: 
IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII
Before you create the issue:
IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII
Search for similar report among other reported issues.
Learn how to troubleshoot at https://www.mautic.org/docs/en/tips/troubleshooting.html
Use drag&drop to attach images or other files )

## Bug Description
Relative date segment filter (like -x days/+5 days) doesn't take into account Mautic timezone setting. It assumes user is in the UTC timezone.
| Q   | A
| --- |
| Mautic version | 2.15.3
| PHP version | 7.2
| Browser | Safari / Any

### Steps to reproduce
1. Create a couple of identified leads, e.g. 6.
2. Go to the `leads` table and update `date_added` value for them as shown on the screenshot.
Snímek obrazovky 2019-10-18 v 19 15 21
3. Create a segment with relative `Date Added` filter to match the majority of leads. E.g. if today is 18th of October and you updated `date_added` as shown on the screenshot above, you should enter `-3 days` to match 15th of October.
Snímek obrazovky 2019-10-18 v 19 54 55
4. Run `php app/console mautic:segments:update` command, go to the `lead_lists_leads` table to check what leads were added to your segment.

### Expected result:
Leads must be added to your segment depending on your local Mautic timezone setting.
E.g. if it set to `Europe/Paris, you should get leads between `2019-10-14 22:00:00` and `2019-10-15 21:59:59` because `2019-10-15 00:00:00` in Paris timezone equals `2019-10-14 22:00:00` UTC.
So the following leads IDS must be added to your segment: 2,3,4,5.

### Actual Result:
The filter doesn't take into account your Mautic timezone settings. You will get leads that have either `2019-10-14%` or `2019-10-15%` as their `date_added` value.
I mean either leads IDs 1, 2 or 3, 4, 5, 6. But most of the times your are not in the UTC timezone and this is not correct.
To fix it segment query filter should get leads using the range date filter, I mean: 
```sql
WHERE `date_added` >= `2019-10-14 22:00:00` AND `date_added` < `2019-10-15 21:59:59`.
``` 
Now it uses
```sql
date_added LIKE `2019-15-10%`
```
and this is nor correct or accurate. It should take into account the actual timezone setting.
Also, https://github.com/mautic/mautic/pull/7752 has to be merged first before fixing this issue . That other PR fixes bug in the DateTimeHelper::setDateTime method.

[//]: # ( Invisible comment:
Please check for related errors in the latest log file in [mautic root]/app/log/ and/or the web server's logs and post them here. Be sure to remove sensitive information if applicable. )

### PR DESCRIPTION
**Pull request by mohit-rocks**
closes https://github.com/mautic/mautic/issues/7967
Rebasing https://github.com/mautic/mautic/pull/8044 against 3.x and resolving merge conflicts.

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? |  Y
| New feature? | 
| Automated tests included? |
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | #8044 #7967
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
1. Create a couple of identified leads, e.g. 6.
2. Go to the `leads` table and update `date_added` value for them as shown on the screenshot.
Snímek obrazovky 2019-10-18 v 19 15 21
3. Create a segment with relative `Date Added` filter to match the majority of leads. E.g. if today is 18th of October and you updated `date_added` as shown on the screenshot above, you should enter `-3 days` to match 15th of October.
Snímek obrazovky 2019-10-18 v 19 54 55
4. Run `php app/console mautic:segments:update` command, go to the `lead_lists_leads` table to check what leads were added to your segment.

### Expected result:
Leads must be added to your segment depending on your local Mautic timezone setting.
E.g. if it set to `Europe/Paris, you should get leads between `2019-10-14 22:00:00` and `2019-10-15 21:59:59` because `2019-10-15 00:00:00` in Paris timezone equals `2019-10-14 22:00:00` UTC.
So the following leads IDS must be added to your segment: 2,3,4,5.

### Actual Result:
The filter doesn't take into account your Mautic timezone settings. You will get leads that have either `2019-10-14%` or `2019-10-15%` as their `date_added` value.
I mean either leads IDs 1, 2 or 3, 4, 5, 6. But most of the times your are not in the UTC timezone and this is not correct.
To fix it segment query filter should get leads using the range date filter, I mean: 
```sql
WHERE `date_added` >= `2019-10-14 22:00:00` AND `date_added` < `2019-10-15 21:59:59`.
``` 
Now it uses
```sql
date_added LIKE `2019-15-10%`
```
and this is nor correct or accurate. It should take into account the actual timezone setting.
Also, https://github.com/mautic/mautic/pull/7752 has to be merged first before fixing this issue . That other PR fixes bug in the DateTimeHelper::setDateTime method.

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Follow the steps above

### CODE REVIEW

**Comment by npracht** [comment]
@mohit-rocks can you check travis please ?

**Comment by mohit-rocks** [comment]
Sure, I'll update it today itself. Thanks for the quick review.

**Comment by npracht** [comment]
@mohit-rocks still an issue with tests

**Comment by RCheesley** [comment]
Open & close for Scrutinizer

**Comment by dennisameling** [comment]
Needs https://github.com/mautic/mautic/pull/7752 to be merged first, will rebase this PR once that's done 👍

**Comment by dennisameling** [comment]
@mohit-rocks this PR has some conflicts now that #7752 is merged. Could you please resolve those so we can go ahead and test this PR? `DateTimeHelper.php` seems to be the culprit here. Thanks!!

**Comment by RCheesley** [comment]
Hi there, we have decided to move this back to 3.0.2 due to the conflicts - if you can address them then we can get it tested and into the release!

**Comment by mohit-rocks** [comment]
@dennisameling  @RCheesley 
I've updated the PR and resolved the conflicts.

**Comment by RCheesley** [comment]
Close & re-open for Travis

File: app/bundles/CoreBundle/Helper/DateTimeHelper.php
Comment (dennisameling) [inline app/bundles/CoreBundle/Helper/DateTimeHelper.php]: Here you're using `$localTimezone`, but a few lines down you're using `$this->localTimezone`. That's inconsistent. From reading the code I think the line here is superfluous.

File: app/bundles/CoreBundle/Helper/DateTimeHelper.php
Comment (dennisameling) [inline app/bundles/CoreBundle/Helper/DateTimeHelper.php]: I think this is causing the tests to fail, because in https://github.com/mautic/mautic/pull/7752/files they're always using `date_default_timezone_get()` but you're taking the `default_timezone` from the config and if there is none, then use `date_default_timezone_get()`

**Comment by RCheesley** [comment]
Bumping to the next release due to pending requests for changes.

**Comment by RCheesley** [comment]
@mohit-rocks having to bump this one again due to unresolved feedback. Please take a look and we can hopefully get this tested and merged in 3.1.1! 🚀

**Comment by RCheesley** [comment]
@mohit-rocks any chance you can take a look at the feedback on the PR? We need that to be addressed if we are to merge this PR. If you could give us a timescale for when you can look into it that would be awesome!

**Comment by mohit-rocks** [comment]
@RCheesley I am reviewing it today, so I'll share some update by tomorrow.

**Comment by mabumusa1** [comment]
Hello @mohit-rocks, please let me know if you need help on this issue

**Comment by RCheesley** [comment]
@mohit-rocks having to bump this to the next release, so we will need to get the issues addressed by **18th October** for inclusion in 3.1.2.

**Comment by RCheesley** [comment]
@cla-bot check

**Comment by dennisameling** [comment]
@mohit-rocks could you please rebase this PR to the `3.1` branch by doing `git rebase upstream/3.1`, then `git push -f`? When that's done I'm happy to re-test this PR. We'll then also see whether or not this PR decreases code coverage. If it decreases the coverage, we'll need to add/update some tests. Thanks in advance 🚀

**Comment by mohit-rocks** [comment]
Thanks @dennisameling. I've rebased PR.

**Comment by dennisameling** [comment]
Thanks, I'll make sure to re-test this (hopefully some time next week) for the 3.2 release as we couldn't squeeze it into 3.1.2 anymore. To be continued 😊

**Comment by RCheesley** [comment]
@mohit-rocks were you able to take on board the comments from @dennisameling above as they seem unresolved!

**Review by pwned555 — APPROVED** [review]
I was able to reproduce the issue in the steps and after applying the code changes I got the desired results in the output.

This change also fixed another issue not discussed in the bug.  Before if I was to try to use a filter like -1 minute it would replace that with an actual date and not be dynamic going forward (the date also wasn't -1 minute).  This now leaves the -1 minute text and works as you would expect.

Finally there were issues with using a filter of today for a datetime.  For example if I used where Expiry date < today, it wasn't adding contacts to the segment until their expiry was before midnight of the current date.
Ex:
User 1 expires March 1st at 9am
User 2 expires March 1st at 10am
User 3 expires March 2nd at 3pm.

The first two users wouldn't get added to the segment until March 2nd at midnight UTC Time.  With this update it now checks the current time as well as date and doesn't use midnight for today.  This means if the segment were to check for updates on March 1st at 9:01 it adds user 1 to the segments (as I would expect) and so on.

**Comment by RCheesley** [comment]
@kuzmany what is blocking this PR? I saw that you added the label but did not add any comment. The PR mentioned that needed to be merged before this one has already been merged, but maybe you are aware of something that I am not?

**Comment by kuzmany** [comment]
@RCheesley 
I forgot write comment, sorry.
We are the author of this PR and this is based on wrong idea. Local timezone means the from local server, not from configuration. Need be rewrite, If this issue still exists, because meanwhile I saw some improvements from Acquia.

**Comment by RCheesley** [comment]
OK - so you mean that if a contact had a timezone set they would not be respected due to the changes in this PR as it is only taking into account the timezone of the Mautic instance?

**Comment by kuzmany** [comment]
@RCheesley this DateTimeHelper use UTC and local. 
Local is taken from server timezone.

If we want to use locale from Configuration, we need read it from parameters and add it to where we need use it (Campaign use it for example).

But this already merged PR do something similar https://github.com/mautic/mautic/pull/7244/files#diff-b732ab5ba852c8f8a8535eb57e71f1cfc342da13dbfbd7e300ccc1e4ed88066cR48

Can you please test it if you have free time if issue still persist?

**Comment by RCheesley** [comment]
@pwned555 were you on 3.3.1 before you applied this PR? If so then the issue seems to still persist.

**Comment by kuzmany** [comment]
I close it favor to https://github.com/mautic/mautic/pull/9784

**Comment by pwned555** [comment]
I was on 3.2.1.  When we upgrade I will test if it's fixed.

**Pull request closed without merging**
